### PR TITLE
Fix server island build error

### PR DESCRIPTION
**Pull request by matthewp**
## Changes

- Fix server island manifest placeholder replacement when generated chunks use double-quoted string literals.
- Use quote-agnostic matching for detection and replacement in server-island build hooks.
- Fixes https://github.com/withastro/astro/issues/15886

## Testing

- Add a regression unit test that simulates quote-style conversion and asserts the output contains materialized `serverIslandMap`/`serverIslandNameMap` maps.

## Docs

- No docs update needed; this is a bug fix for incorrect runtime behavior.

### CODE REVIEW

File: packages/astro/src/core/server-islands/vite-plugin-server-islands.ts [hunk 1]
```diff
@@ -10,8 +10,11 @@ import { ASTRO_VITE_ENVIRONMENT_NAMES } from '../constants.js';
 export const SERVER_ISLAND_MANIFEST = 'virtual:astro:server-island-manifest';
 const RESOLVED_SERVER_ISLAND_MANIFEST = '\0' + SERVER_ISLAND_MANIFEST;
 
-export const serverIslandPlaceholderMap = "'$$server-islands-map$$'";
+const serverIslandPlaceholderMap = "'$$server-islands-map$$'";
```
Comment (ematipico) [inline packages/astro/src/core/server-islands/vite-plugin-server-islands.ts:10]: If you're looking for a solution that is vite 8 friendly, use template literals.

Turns out that oxc/rolldown do a bunch of normalisation with strings, and sometimes it transforms them

**Comment by CaiJimmy** [comment]
Thanks for the quick fix! 

I have encountered now with a different error in preview environment (using the same reproducible example, but with astro package given by pkg-pr-new):

```
GET / 200 OK (5ms)
GET /.well-known/appspecific/com.chrome.devtools.json 404 Not Found (3ms)
19:43:32 [ERROR] Error: No such module "dist/server/chunks/Island_!~{00b}~.mjs".
    at file:///Users/jimmy/Documents/repro-cf-v6/dist/server/chunks/_virtual_astro_server-island-manifest_zfG9aQW0.mjs:4:19
    at page (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:9222:35)
    at async callComponentAsTemplateResultOrResponse (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:7116:25)
    at async renderToReadableStream (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:7064:26)
    at async renderPage (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:8320:14)
    at async lastNext (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:11522:25)
    at async callMiddleware (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:8412:10)
    at async RenderContext.render (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:11556:73)
    at async App.render (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:12451:20)
    at async Object.handle [as fetch] (file:///Users/jimmy/Documents/repro-cf-v6/dist/server/entry.mjs:12961:20)
GET /_server-islands/Island 500 Internal Server Error (4ms)
```

image

It's trying to import a non-existent file from chunks folder. The dev server does not work neither. Not sure if this is related. Should I create a separate issue?

File: packages/astro/src/core/server-islands/vite-plugin-server-islands.ts (hunk 1 above)
Comment (matthewp) [reply packages/astro/src/core/server-islands/vite-plugin-server-islands.ts:10]: omg

**Comment by matthewp** [comment]
@CaiJimmy no, will keep this one.

File: packages/astro/src/core/server-islands/vite-plugin-server-islands.ts (hunk 1 above)
Comment (matthewp) [reply packages/astro/src/core/server-islands/vite-plugin-server-islands.ts:10]: Oh wait, that doesn't matter here, this code isn't loaded by Vite.

**Comment by CaiJimmy** [comment]
Thanks!

I just tested with the current stable version of Astro 6 and can confirm that it does not generate the `dist/server/chunks/Island_!~{00b}~.mjs` file neither.

**Comment by matthewp** [comment]
@CaiJimmy can you try with:

```shell
npm i https://pkg.pr.new/astro@fbf9d83
```

**Comment by CaiJimmy** [comment]
@matthewp I can confirm that now preview environment working correctly with `server:defer`, generating the correct chunk file.

But `dev` server still gives error with `prerenderEnvironment: 'node'`. I can see in the console that it's a 400 error, but no error log is given currently so I'm afraid that I cannot provide any more useful information to you:

```
➜  repro-cf-v6 git:(main) ✗ pnpm dev

> repro-cf-v6@0.0.1 dev /Users/jimmy/Documents/repro-cf-v6
> astro dev

20:45:46 [@astrojs/cloudflare] Enabling image processing with Cloudflare Images for production with the "IMAGES" Images binding.
20:45:46 [@astrojs/cloudflare] Enabling sessions with Cloudflare KV with the "SESSION" KV binding.
[vite] connected.
20:45:47 [types] Generated 0ms
20:45:47 [WARN] [content] Content config not loaded
 astro  v6.0.4 ready in 1155 ms
┃ Local    http://localhost:4321/
┃ Network  use --host to expose
20:45:47 watching for file changes...
20:45:48 [200] / 17ms
20:45:48 [200] / 20ms
20:45:48 [vite] 20:45:48 [404] /@id/astro/runtime/client/dev-toolbar/astro_runtime_client_dev-toolbar_entrypoint__js.js.map 9ms
20:45:48 [vite] 20:45:48 [400] /_server-islands/Island 12ms
```

Is there any way to expose more error stack information here?

**Comment by CaiJimmy** [comment]
With `prerenderEnvironment: 'workerd'`, server island in `dev` server seems to be working correctly.

**Comment by matthewp** [comment]
@CaiJimmy I think I know the dev issue here. It's unrelated to this change so I'll look into that as a separate PR.

**Comment by matthewp** [comment]
@CaiJimmy https://github.com/withastro/astro/pull/15891

**Comment by CaiJimmy** [comment]
Thank you!

I'm currently testing this new version on my website, and found that there is some issue with importing third party libraries (using the top level `import`, not `require`)

The example server island would be something like this:

```astro
---
import { encodeBase64 } from "@oslojs/encoding";
 
const data: Uint8Array = new TextEncoder().encode("hello world");
const encoded = encodeBase64(data);
---

<p>Encoded: {encoded}</p>
```

The resulting file after running build keeps `import { encodeBase64 } from "@oslojs/encoding";`, and results error in `preview` (and I guess also in the real cloudflare worker environment):



Inside `./transition_BPFN3ezl.mjs`, there is another direct import from the library:

```
import { decodeBase64, encodeBase64, decodeHex, encodeHexUpperCase } from '@oslojs/encoding';
```

Error when running preview:

```
➜  repro-cf-v6 git:(main) pnpm preview

> repro-cf-v6@0.0.1 preview /Users/jimmy/Documents/repro-cf-v6
> astro preview

21:27:44 [@astrojs/cloudflare] Enabling image processing with Cloudflare Images for production with the "IMAGES" Images binding.
21:27:44 [@astrojs/cloudflare] Enabling sessions with Cloudflare KV with the "SESSION" KV binding.
21:27:44 [@astrojs/cloudflare]
 astro  v13.1.1 ready in 29 ms

┃ Local    http://localhost:4321/

GET / 200 OK (14ms)
21:27:46 [ERROR] Error: No such module "dist/server/chunks/@oslojs/encoding".
  imported from "dist/server/chunks/Island_CZSxsmqZ.mjs"
GET /_server-islands/Island 500 Internal Server Error (17ms)
21:27:46 [ERROR] Error: No such module "dist/server/chunks/@oslojs/encoding".
  imported from "dist/server/chunks/Island_CZSxsmqZ.mjs"
GET /_server-islands/Island 500 Internal Server Error (2ms)
```

I believe (?) this is not related with https://docs.astro.build/en/guides/integrations-guide/cloudflare/#some-dependencies-might-need-to-be-pre-compiled, because it's using the ES6 export and import syntax.

I have updated the repo with a minimal reproducible example: https://github.com/CaiJimmy/astro-v6-prerender

Should I create a separate issue for this? Seems like a bundle problem.

**Comment by matthewp** [comment]
@CaiJimmy No, I think that might be related to this fix. I'll try and get another test for this scenario and it fixed here.

File: packages/astro/src/core/server-islands/vite-plugin-server-islands.ts (hunk 1 above)
Comment (ematipico) [reply packages/astro/src/core/server-islands/vite-plugin-server-islands.ts:10]: It actually does... 😅 

I had to apply this change https://github.com/withastro/astro/pull/15819/changes/1cae17182134e56f18fe93856ca97148f6bcafc2

It's just a FYI, no need to change it here. I wanted to let you know

**Comment by matthewp** [comment]
@CaiJimmy here's new ones for you to try:

```
npm i https://pkg.pr.new/astro@a68cdfa
npm i https://pkg.pr.new/@astrojs/cloudflare@a68cdfa
```

**Comment by CaiJimmy** [comment]
Thanks!

With `prerenderEnvironment` set to `workerd`, third party libraries are correctly bundled:

```js
/// island_CllqxkrA.mjs
globalThis.process ??= {};
globalThis.process.env ??= {};
import { A as AstroError, ao as InvalidComponentArgs, ap as encodeBase64, l as maybeRenderHead, D as renderTemplate } from "./transition_BffVS1kO.mjs";
```

However, when it's set to `node`, it will no longer bundle:

```js
/// island_C0XpBDl0.mjs
globalThis.process ??= {}; globalThis.process.env ??= {};
import { A as AstroError, ao as InvalidComponentArgs, l as maybeRenderHead, D as renderTemplate } from './transition_BPFN3ezl.mjs';
import { encodeBase64 } from '@oslojs/encoding';
```

The test was done using this example: https://github.com/CaiJimmy/astro-v6-prerender

**Comment by matthewp** [comment]
@CaiJimmy What is incorrect about the second example there? @oslo/encoding is not bundled because dependencies are not bundled by default in Node.js. Does this cause a problem?

**Comment by CaiJimmy** [comment]
> @CaiJimmy What is incorrect about the second example there? @oslo/encoding is not bundled because dependencies are not bundled by default in Node.js. Does this cause a problem?

Yes, because this dependency is used by the Server Island script, which will be executed in the Worker environment.

It wouldn't be a problem if the pre-render page itself used it because it would be executed in a normal Node environment.

**Comment by matthewp** [comment]
Right ok, that's interesting. You are hitting the hardest part of this refactor 😅. I've been trying to avoid flipping the order that the environments run, but I might need to do that. Let me try it out.

**Comment by matthewp** [comment]
@CaiJimmy can you try this one?

```
npm i https://pkg.pr.new/@astrojs/cloudflare@2cb31aa
npm i https://pkg.pr.new/astro@2cb31aa
```

**Comment by CaiJimmy** [comment]
Thanks!

I have tested it, and it is now compiling and bundling the dependencies correctly.

I have also tested this version on my website and found that Server Island requests are returning empty responses in `preview` server. I need some time to create a minimal reproducible example, but I think it's probably related to the combination of Server Island and the front-end framework hydration (I'm using Solid.js), because plain text inside Server Island is returned in the response correctly.

Thanks again for the fix!

Edit: here is the minimal reproducible example: https://github.com/CaiJimmy/astro-v6-prerender/blob/main/src/pages/island.astro

Solid.js component (I haven't tested yet other frameworks) inside a Astro Server Island component is not rendered:

```astro
---
import A from "./A";
---

<A content={'test'} />
```

```jsx
export default function A({ content }: { content?: string }) {
    return (
        <div>
            <p>Content: {content}</p>
        </div>
    );
}
```

Empty response is returned by the server (it's a 200 response, no error in console):

image

Edit2: I have tested with React, and it has the same issue.
Edit3: This issue is present in both node and workerd prerender environment

**Comment by soklovsk** [comment]
Fixes [15903](https://github.com/withastro/astro/issues/15903).

**Comment by matthewp** [comment]
@CaiJimmy I think I'm going to release this as I'm confident it fixes most of the server island issues people are facing.  Once that's out if you could file a new issue about nested client island inside I'll get that resolved separately.
